### DISCUSSION TITLE
Attribute [ArrayBoundsCheck (false) ] in v168

### DISCUSSION THREAD

**Discussion by dmrvb** [twinbasic/twinbasic#1331] [Q&A]
Exciting stuff, I am trying to assess this.

It appears that this will only work in tB modules, not in classes imported from VB, is that correct?

How do I make this work at procedure level ? I get the error 'unrecognized symbol '

Why do we need two syntaxes for Attributes:
Attribute VB_Name = "FFTd"
[ ArrayBoundsChecks (False) ]

would it be neater to combine them (this might have been discussed before?)

**Comment by dmrvb** [comment]
I see what I was doing wrong,  the [ArrayBoundsChecks (false) ] has to be immediately above a procedure definition. (after any module level variable definitions for the first procedure)

It looks like putting the attribute right above a procedure applies to just that procedure, how would I make it apply to the entire component?

**Reply by WaynePhillipsEA** [reply]
Yeah that's right;. tB style attributes must always appear above the declaration.  As such, if you're using BAS/CLS files, you can't apply tB attributes to the whole component as the attribute has to appear above the outer Module/Class definition, which doesn't explicitly exist for the legacy BAS/CLS files.

You'd need to upgrade those legacy files to TWIN files, and there you can use the component scope attributes.

```
[ ArrayBoundsChecks(False) ]
Class Foo
End Class
```

**Comment by dmrvb** [comment]
Thanks,
I do like fast code, and with a bit of care VB can do reasonably fast code.
Here are some timings for running a loop of 100 65k FFT's (Fast Fourier Transforms)

VB (optimised for fast code, no advanced options)   0.965 sec
tB                                                                                1.27 sec

VB (fast, plus arraybounds and integer overflow removed)  0.79 sec
tB                                                                                            0.83

VB   all advanced optimisations selected                              0.152   

tB is doing pretty well. I will have a play to see which additional VB optimisation is getting that huge speed gain

**Reply by WaynePhillipsEA** [reply]
Bear in mind that tB still produces completely unoptimized compiled code (closer to VB6 pcode than native code).  These attributes just make the compiled code a bit leaner.

Once we start using LLVM for the codegen, I am confident we will surge past VB6 in all cases.

**Reply by dmrvb** [reply]
Yes, but at 0.83 to 0.79 tB is already doing very well even without LLVM.

**Comment by Kr00l** [comment]
It would be nice to have a documentation about the whole attributes topic.
It's getting more and more and having a list of all available attributes for each "place" would be helpful.

**Reply by WaynePhillipsEA** [reply]
Completely agree.  Just a simple list for each scope would be better than nothing.  I'll see what I can do tomorrow.

As discussed previously, we will almost definitely want a better UX for entering/editing attributes, but having a reference for now would be good.

**Comment by dmrvb** [comment]
Selecting the VB compile optimisation "remove floating point error checks" gets the big speed gain. Could this be easily provided as another attribute in tB?

**Reply by WaynePhillipsEA** [reply]
Yes, that could be offered similarly.  Watch this space.

**Reply by WaynePhillipsEA** [reply]
`[ FloatingPointErrorChecks(False) ]` is now available in BETA 169.  I expect you won't see such a dramatic difference in the tB version until the LLVM optimization is in place to take full advantage of being able to combine and mutate the FPU instructions more freely with the floating point checks off.  Regardless, I'd be interested to hear what difference it makes in your case.

**Comment by dmrvb** [comment]
You are correct, removing the floating point error checks makes no measurable difference to execution time. I wonder why it has such a big impact in VB6 code????
The mousepointer now displays the correct "hourglass" whilst the code is running, thankyou.

I believe that the [ checkstuff (false) ] should go after rather than before the "Public Sub DavesCode()".
The Sub and End Sub are effectively defining a block of code and everything relating to that code should be contained within them.
I put any procedure description comments within the procedure whilst many programmers put them above it.
I am probably in a minority of one here, but I thought I'd say it anyway. :-)

**Reply by WaynePhillipsEA** [reply]
Removing the intermediate error checks allows an optimizer to do much more aggressive optimizing.  Complex floating point expressions in particular can often be heavily optimized by a good compiler.  We don't have an optimizing compiler yet, but when we do, you can be sure this new attribute will be more effective and more in line with the VB6 native code compiler.

**Reply by WaynePhillipsEA** [reply]
With regards attributes, the syntax and placement was discussed at length in https://github.com/twinbasic/twinbasic/issues/835.

As you'll note in that (long) issue, it was not possible for the community at the time to come to a conclusive agreement.  However, decisions had to be made, and it was felt that keeping attributes always immediately above declarations was clear and keeps the code itself clean.  

Additionally, in future, the IDE will likely not show the attributes by default, with them being accessible via some sort of expanding node or property sheet, so the actual stored placement of the attributes within the file will become less of an issue.

**Reply by fafalone** [reply]
There's precedent for this; in VB6, there's attributes in classes, that are hidden from the user but set with the 'Procedure Attributes' dialog under Tools.  Those are placed above. So it was right to keep that design in line with VB, since presumably that placement has to be supported anyway.

**Comment by dmrvb** [comment]
Not sure about that, I would prefer everything to be visible, having things like this hidden could but be a source of hard to spot bugs?

**Reply by WaynePhillipsEA** [reply]
All discussed in the linked issue.  Some people want them to be always visible.  Some don't.  Don't worry, you'll always be given a choice with tB.  

(ps I'll personally always want them visible too)
